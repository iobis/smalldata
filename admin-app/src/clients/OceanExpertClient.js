export function getByExpertByName(name) {
  const defaultResult = []
  if (!name || !name.trim()) return defaultResult
  const url = 'https://www.oceanexpert.net/api/v1/advanceSearch/search.json?action=browse&type=all&query=' + name.trim()
  return fetch(url)
    .then(response => response.json())
    .then(json => (json.results.data || []).filter(record => !!record.name && record.type === 'experts'))
}
