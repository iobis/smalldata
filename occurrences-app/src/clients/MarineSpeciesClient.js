export function getByName(scientificName) {
  const url = 'http://www.marinespecies.org/rest/AphiaRecordsByName/' + scientificName + '?like=true&marine_only=false'
  return fetch(url).then(response => response.json())
}

export function getById(scientificNameId) {
  const id = scientificNameId.split(':').pop()
  const url = 'http://www.marinespecies.org/rest/AphiaRecordByAphiaID/' + id
  return fetch(url).then(response => response.json())
}

export function isScientificNameId(param) {
  return !!param && param.split(':').length > 0 && /^\d+$/.test(param.split(':').pop())
}
