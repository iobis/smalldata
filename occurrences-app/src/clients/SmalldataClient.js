import { format } from 'date-fns'

export async function getDatasets() {
  const response = await fetch('/api/datasets')
    .then(response => response.json())
  return response.map(renameRefToId)
}

export function renameRefToId({ ref, ...rest }) {
  return ({ id: ref, ...rest })
}

export function datasetTitleOf(dataset) {
  const title = dataset && dataset.title && dataset.title.value
  return title || ''
}

export function getOccurrenceMock() {
  return [{
    id:             1,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }, {
    id:             2,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }]
}

export async function postOccurrence({ occurence }) {
  const userRef = 'ovZTtaOJZ98xDDY'
  const occurrence = {
    core:       'occurrence',
    occurrence: [{
      tdwg: {
        datasetID:                     occurence.dataset.id,
        datasetName:                   occurence.dataset.title.value,
        basisOfRecord:                 occurence.occurrenceData.basisOfRecord.charAt(0).toUpperCase() + occurence.occurrenceData.basisOfRecord.slice(1),
        eventDate:                     format(occurence.occurrenceData.beginDate, 'YYYY-MM-DD') + '/' + format(occurence.occurrenceData.endDate, 'YYYY-MM-DD'),
        lifestage:                     occurence.occurrenceData.lifeStage,
        occurrenceStatus:              occurence.occurrenceData.occurrenceStatus,
        scientificName:                occurence.occurrenceData.scientificName,
        sex:                           occurence.occurrenceData.sex,
        decimalLongitude:              occurence.locationData.decimalLongitude,
        decimalLatitude:               occurence.locationData.decimalLatitude,
        coordinateUncertaintyInMeters: occurence.locationData.coordinateUncertainty,
        minimumDepthInMeters:          occurence.locationData.minimumDepth,
        maximumDepthInMeters:          occurence.locationData.maximumDepth,
        verbatimCoordinates:           occurence.locationData.verbatimCoordinates,
        verbatimDepth:                 occurence.locationData.verbatimDepth
      }
    }],
    emof:       [{
      purl:  {},
      iobis: {}
    }, {
      iobis: {}
    }]
  }
  const url = `/api/dwca/${occurence.dataset.id}/user/${userRef}/records`
  return await fetch(url, {
    method:  'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body:    JSON.stringify(occurrence)
  }).then(response => response.json())
}
