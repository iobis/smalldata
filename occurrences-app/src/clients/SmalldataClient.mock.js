export const DATASTES_RESPONSE = [{
  'meta':              { 'type': 'event', 'dwcTables': { 'core': 'event', 'extensions': ['emof', 'occurrence'] } },
  'title':             {
    'language': 'en',
    'value':    'Caprellids polulation structure in Usujiri, Hokkaido, Japan'
  },
  'language':          'en',
  'abstract':          { 'paragraphs': ['Population structure data of Caprella species used in the article "Hosono T (2014) Temperature explains reproductive dynamics in caprellids at different latitudes. Marine Ecology Progress Series 511:129-141"'] },
  'license':           {
    'url':   'http://creativecommons.org/licenses/by-nc/4.0/legalcode',
    'title': 'Creative Commons Attribution Non Commercial (CC-BY-NC) 4.0 License'
  },
  'creators':          [{ 'individualName': { 'givenName': 'Takashi', 'surName': 'Hosono' } }],
  'contacts':          [{ 'individualName': { 'givenName': 'Takashi', 'surName': 'Hosono' } }],
  'metadataProviders': [{
    'individualName':        { 'givenName': 'Takashi', 'surName': 'Hosono' },
    'electronicMailAddress': 'hosonot@jamstec.go.jp'
  }],
  'keywordSets':       [{
    'keywords':         ['Samplingevent'],
    'keywordThesaurus': 'GBIF Dataset Type Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_type.xml'
  }],
  'ref':               'wEaBfmFyQhYCdsk'
}, {
  'meta':              { 'type': 'occurrence', 'dwcTables': { 'core': 'occurrence', 'extensions': ['emof'] } },
  'title':             {
    'language': 'en',
    'value':    'Benthos collected in the Azov Sea in 1935 on board the R/V "N. Danilevskiy"'
  },
  'language':          'en',
  'abstract':          { 'paragraphs': ['The data were collected in the Azov Sea during a cruise on board of R/V N. Danilevskiy in 1935. This dataset contains abundance data (individuals per m²) and biomass data (g/m²) for benthic taxa. No additional metadata is available.', 'This dataset was rescued through data archaeology activities in the framework of the EMODnet Biology data grants.'] },
  'license':           {
    'url':   'http://creativecommons.org/licenses/by/4.0/legalcode',
    'title': 'Creative Commons Attribution Non Commercial (CC-BY) 4.0 License'
  },
  'creators':          [{
    'individualName':        { 'givenName': 'Yu. B.', 'surName': 'Belogurova' },
    'organizationName':      'National Academy of Sciences of Ukraine Institute of Biology of the Southern Seas (NASU-IBSS)',
    'electronicMailAddress': 'director@ibss.iuf.net'
  }, {
    'individualName':        { 'givenName': 'Oleksandra', 'surName': 'Sergeyeva' },
    'organizationName':      'National Academy of Sciences of Ukraine Institute of Biology of the Southern Seas (NASU-IBSS)',
    'electronicMailAddress': 'o.sergeyeva@gsva.net'
  }],
  'contacts':          [{
    'individualName':        { 'givenName': 'Oleksandra', 'surName': 'Sergeyeva' },
    'organizationName':      'National Academy of Sciences of Ukraine Institute of Biology of the Southern Seas (NASU-IBSS)',
    'electronicMailAddress': 'o.sergeyeva@gsva.net'
  }],
  'metadataProviders': [{
    'individualName':        { 'surName': 'EurOBIS Data Management Team' },
    'organizationName':      'Flanders Marine Institute (VLIZ)',
    'electronicMailAddress': 'info@eurobis.org'
  }],
  'keywordSets':       [{
    'keywords':         ['Occurrence'],
    'keywordThesaurus': 'GBIF Dataset Type Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_type.xml'
  }, {
    'keywords':         ['Observation'],
    'keywordThesaurus': 'GBIF Dataset Subtype Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_subtype.xml'
  }],
  'ref':               'ntDOtUc7XsRrIus'
}, {
  'meta':              { 'type': 'occurrence', 'dwcTables': { 'core': 'occurrence', 'extensions': ['emof'] } },
  'title':             {
    'language': 'en',
    'value':    'Antipatharia distribution data from: Deep-sea fauna of European seas - an annotated species check-list of benthic invertebrates living deeper than 2000 m in the seas bordering Europe'
  },
  'language':          'en',
  'abstract':          { 'paragraphs': ['An annotated check-list is given of Antipatharia species occurring deeper than 2000 m in the seas bordering Europe.', 'The check-list is based on published data. It includes 12 species. For each species synonymy, data on localities in European seas and general species distribution are provided. Only distributions for which geographic coordinates are availlable were retained, as these are made available through EurOBIS.'] },
  'license':           {
    'url':   'http://creativecommons.org/licenses/by/4.0/legalcode',
    'title': 'Creative Commons Attribution Non Commercial (CC-BY) 4.0 License'
  },
  'creators':          [{
    'individualName':        { 'givenName': 'Tina', 'surName': 'Molodtsova' },
    'organizationName':      'Russian Academy of Sciences P. P. Shirshov Institute of Oceanology (RAS-SIO RAS)',
    'electronicMailAddress': 'tina@ocean.ru'
  }],
  'contacts':          [{
    'individualName':        { 'givenName': 'Tina', 'surName': 'Molodtsova' },
    'organizationName':      'Russian Academy of Sciences P. P. Shirshov Institute of Oceanology (RAS-SIO RAS)',
    'electronicMailAddress': 'tina@ocean.ru'
  }],
  'metadataProviders': [{
    'individualName':        { 'surName': 'EurOBIS Data Management Team' },
    'organizationName':      'Flanders Marine Institute (VLIZ)',
    'electronicMailAddress': 'info@eurobis.org'
  }],
  'keywordSets':       [{
    'keywords':         ['Occurrence'],
    'keywordThesaurus': 'GBIF Dataset Type Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_type.xml'
  }, {
    'keywords':         ['Observation'],
    'keywordThesaurus': 'GBIF Dataset Subtype Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_subtype.xml'
  }],
  'ref':               'PoJnGNMaxsupE4w'
}, {
  'meta':              { 'type': 'occurrence', 'dwcTables': { 'core': 'occurrence', 'extensions': ['emof'] } },
  'title':             { 'language': 'en', 'value': 'Benthic data from Sevastopol (Black Sea)' },
  'language':          'en',
  'abstract':          {
    'paragraphs': ['The dataset contains the abundance and biomass values of sediment organisms from the Sevastopol region (South-Western Crimea, The Black Sea, Ukraine) collected between 1982 and 1992.', 'Abundance and Biomass of benthic organisms and key abiotic variables (covering irregular sampling surveys within 1982-1992) from the Sevastopol Area (South-Western Crimea, The Black Sea, Ukraine) including estuaries and small sheltered and semi-closed bays within the area of &quot;big Sevastopol bay&quot;\nSevastopol Bay is located in south western part of the Crimean peninsula, Ukraine and comprehends the one of the biggest marine harbours of the Black Sea and has served as an industrialized naval base for decades. A consequence of this long-term industrial, military and municipal activity was the deterioration of the marine environment of the bay, depressed state of benthic assemblages and occurrence of expanded bottom areas that became contaminated with various sediment-associated pollutants including polycyclic aromatic hydrocarbons (PAHs), polychlorinated biphenyls (PCBs), pesticides, and metals. Benthic samples in Sevastopol bay heve been taken between 1982 and 1992.']
  },
  'license':           {
    'url':   'http://creativecommons.org/licenses/by-nc/4.0/legalcode',
    'title': 'Creative Commons Attribution Non Commercial (CC-BY-NC) 4.0 License'
  },
  'creators':          [{
    'individualName':        { 'givenName': 'Christos', 'surName': 'Arvanitidis' },
    'organizationName':      'Hellenic Centre for Marine Research (HCMR)',
    'electronicMailAddress': 'arvanitidis@her.hcmr.gr'
  }, { 'organizationName': 'Hellenic Centre for Marine Research (HCMR)' }, {
    'individualName':        {
      'givenName': 'Alexei',
      'surName':   'Petrov'
    },
    'organizationName':      'National Academy of Sciences of Ukraine Institute of Biology of the Southern Seas (NASU-IBSS',
    'electronicMailAddress': 'alexpet@ibss.iuf.net'
  }],
  'contacts':          [{
    'individualName':        { 'givenName': 'Christos', 'surName': 'Arvanitidis' },
    'organizationName':      'Hellenic Centre for Marine Research (HCMR)',
    'electronicMailAddress': 'arvanitidis@her.hcmr.gr'
  }],
  'metadataProviders': [{
    'individualName':        { 'surName': 'EurOBIS Data Management Team' },
    'organizationName':      'Flanders Marine Institute (VLIZ)',
    'electronicMailAddress': 'info@eurobis.org'
  }],
  'keywordSets':       [{
    'keywords':         ['Occurrence'],
    'keywordThesaurus': 'GBIF Dataset Type Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_type.xml'
  }, {
    'keywords':         ['Observation'],
    'keywordThesaurus': 'GBIF Dataset Subtype Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_subtype.xml'
  }, { 'keywords': ['Benthic biomass', 'Benthos', 'Data', 'Marine Genomics'], 'keywordThesaurus': 'ASFA' }],
  'ref':               'NnqVLwIyPn-nRkc'
}]

export const OCCURRENCES_RESPONSE = [
  {
    'dwcaId':     'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_796',
    'dataset':    'ntDOtUc7XsRrIus',
    'dateAdded':  null,
    'dwcRecords': {
      'emof':       [
        {
          'id':   'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_796',
          'tdwg': {}
        },
        {
          'id':   'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_796',
          'tdwg': {}
        }
      ],
      'core':       'occurrence',
      'occurrence': [
        {
          'id':   'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_796',
          'tdwg': {
            'scientificName': 'Nereis succinea',
            'eventDate':      '1935-10-02'
          }
        }
      ]
    }
  },
  {
    'dwcaId':     'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_1158',
    'dataset':    'ntDOtUc7XsRrIus',
    'dateAdded':  null,
    'dwcRecords': {
      'emof':       [
        {
          'id':   'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_1158',
          'tdwg': {}
        },
        {
          'id':   'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_1158',
          'tdwg': {}
        }
      ],
      'core':       'occurrence',
      'occurrence': [
        {
          'id':   'IBSS_R/V N. Danilevskiy 1935 Azov Sea benthos data_1158',
          'tdwg': {
            'scientificName': 'Brachynotus lucasi',
            'eventDate':      '1971'
          }
        }
      ]
    }
  },
  {
    'dwcaId':     'DeepSea_Antipatharia_35',
    'dataset':    'PoJnGNMaxsupE4w',
    'dateAdded':  null,
    'dwcRecords': {
      'core':       'occurrence',
      'occurrence': [
        {
          'id':   'DeepSea_Antipatharia_35',
          'tdwg': {
            'scientificName': 'Abyssopathes lyra'
          }
        }
      ]
    }
  }
]
