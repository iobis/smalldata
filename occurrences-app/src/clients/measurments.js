export function getGeneralMeasurements() {
  return [{
    type:   'Pressure',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/PRESPS02/',
    units:  [{ name: 'Kilogram', id: 'http://vocab.nerc.ac.uk/collection/P06/current/KGXX' }]
  }, {
    type:   'Temperature',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/TEMPCU01/',
    units:  [{ name: 'Degrees Celsius', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UPAA' }]
  }, {
    type:   'Salinity',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/PSALCU01/',
    units:  [{ name: 'Degrees Celsius', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UPAA' }]
  }]
}

export function getSpecificMeasurements() {
  return [{
    type:   'ObservedIndividualCount',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/OCOUNT01',
    units:  [{ name: 'Kilogram', id: 'http://vocab.nerc.ac.uk/collection/P06/current/KGXX' }]
  }, {
    type:   'Abundance category of biological entity specified elsewhere',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL06',
    units:  [{ name: 'Millimeter', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UXMM' }]
  }, {
    type:   'Abundance of biological entity specified elsewhere per unit area of the bed',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL02',
    units:  [{ name: 'Meter', id: 'http://vocab.nerc.ac.uk/collection/P06/current/ULAA/' }]
  }, {
    type:   'Abundance of biological entity specified elsewhere per unit volume of the sediment',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/USPBIOSV',
    units:  [{ name: 'Grams', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UGRM' }]
  }, {
    type:   'Abundance of biological entity specified elsewhere per unit volume of the water body',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL01',
    units:  [{ name: 'Micrometer', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UMIC' }]
  }, {
    type:   'Ash-free dry weight biomass (as carbon) of biological entity specified elsewhere per unit volume of the sediment',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/USPBICBM',
    units:  [{ name: 'Grams per square meter', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UGMS' }]
  }, {
    type:   'Ash-free dry weight biomass of biological entity specified elsewhere per unit area of the bed',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL03',
    units:  [{ name: 'Number per square meter', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UPMS' }]
  }, {
    type:   'Ash-free dry weight biomass of biological entity specified elsewhere per unit volume of the water body',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL07',
    units:  [{ name: 'Number per liter', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UCML' }]
  }, {
    type:   'Length of biological entity specified elsewhere',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/OBSINDLX/',
    units:  [{ name: 'Square meter', id: 'http://vocab.nerc.ac.uk/collection/P06/current/UMSQ' }]
  }, {
    type:   'Specimen weight of biological entity specified elsewhere',
    typeId: 'http://vocab.nerc.ac.uk/collection/P01/current/SPWGXX01/',
    units:  [{ name: 'Cubic metres', id: 'http://vocab.nerc.ac.uk/collection/P06/current/MCUB/' }]
  }]
}
