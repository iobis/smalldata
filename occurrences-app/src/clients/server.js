export function getOccurrenceMock() {
  return [{
    id:             1,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }, {
    id:             2,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }]
}
