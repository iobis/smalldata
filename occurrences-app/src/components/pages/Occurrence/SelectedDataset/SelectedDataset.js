import PropTypes from 'prop-types'
import React, { useState } from 'react'

export default function SelectedDataset() {
  const datasetOptions = [{
    id:          0,
    description: 'HAB Region 2: Occurrences of harmful (toxic) algal taxa within an area of interest to El Salvador compiled as part of a literature search project.'
  }, {
    id:          1,
    description: 'NPPSD Short-tailed Albatross Sightings'
  }, {
    id:          2,
    description: 'PANGAEA - Data from Christian-Albrechts-University Kiel'
  }, {
    id:          3,
    description: 'NSIS: List of marine benthic algae from Magdalen Islands, Quebec as recorded in 1979'
  }, {
    id:          4,
    description: 'Seguimiento de 10 cr√≠as de tortuga boba nacidas en 2016 en el litoral valenciano, en el marco del Proyecto LIFE 15 IPE ES 012 (aggregated per 1-degree cell)'
  }, {
    id:          5,
    description: 'Waved Albatross Tracking (aggregated per 1-degree cell)'
  }]
  const [selectedDatasetId, setSeletedDatasetId] = useState(datasetOptions[0].id)

  return (
    <div className="selected-dataset is-fluid">
      <div className="columns">
        <div className="column">
          <table className="table is-striped is-fullwidth">
            <tbody>
            {datasetOptions.map(props => (
              <DatasetOption
                key={props.id}
                onClick={() => setSeletedDatasetId(props.id)}
                checked={props.id === selectedDatasetId}
                {...props}
              />))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  )
}

function DatasetOption({ checked, description, onClick }) {
  return (
    <tr>
      <td><input checked={checked} type="radio" name="dataset" onClick={onClick}/></td>
      <td>{description}</td>
    </tr>
  )
}

DatasetOption.propTypes = {
  checked:     PropTypes.bool.isRequired,
  description: PropTypes.string.isRequired,
  onClick:     PropTypes.func.isRequired
}
