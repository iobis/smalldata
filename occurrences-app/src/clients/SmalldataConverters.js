import { findTypeAndUnitIdByNames, findUnitsByTypeId } from './measurments'
import { format } from 'date-fns'

const purlUrl = 'http://purl.org/dc/terms/'
const tdwgUrl = 'http://rs.tdwg.org/dwc/terms/'
const iobisUrl = 'http://rs.iobis.org/obis/terms/'

export function datasetTitleOf(dataset) {
  const title = dataset && dataset.title && dataset.title.value
  return title || ''
}

export function mapOccurrenceToDwca(occurrence) {
  const emof = occurrence.measurements.map(measurment => {
    const { typeId, unitId } = findTypeAndUnitIdByNames(measurment.type, measurment.unit)
    return {
      tdwg:  {
        measurementType:  measurment.type,
        measurementUnit:  measurment.unit,
        measurementValue: measurment.value
      },
      iobis: {
        measurementTypeID: typeId,
        measurementUnitID: unitId
      }
    }
  })
  const darwinCoreFields = mapDarwinCoreFieldsToRequest(occurrence.darwinCoreFields)
  return {
    core:       'occurrence',
    occurrence: [{
      tdwg:  {
        ...mapOccurrenceDataToTdwg(occurrence.occurrenceData),

        decimalLongitude:              occurrence.locationData.decimalLongitude,
        decimalLatitude:               occurrence.locationData.decimalLatitude,
        coordinateUncertaintyInMeters: occurrence.locationData.coordinateUncertainty,
        minimumDepthInMeters:          occurrence.locationData.minimumDepth,
        maximumDepthInMeters:          occurrence.locationData.maximumDepth,
        verbatimCoordinates:           occurrence.locationData.verbatimCoordinates,
        verbatimDepth:                 occurrence.locationData.verbatimDepth,

        institutionCode:         occurrence.observationData.institutionCode,
        collectionCode:          occurrence.observationData.collectionCode,
        fieldNumber:             occurrence.observationData.fieldNumber,
        catalogNumber:           occurrence.observationData.catalogNumber,
        recordNumber:            occurrence.observationData.recordNumber,
        identifiedBy:            occurrence.observationData.identifiedBy.join('|'),
        recordedBy:              occurrence.observationData.recordedBy.join('|'),
        identificationQualifier: occurrence.observationData.identificationQualifier,
        identificationRemarks:   occurrence.observationData.identificationRemarks,
        associatedReferences:    occurrence.observationData.references.join('|'),

        ...darwinCoreFields.tdwg
      },
      purl:  {
        ...darwinCoreFields.purl
      },
      iobis: darwinCoreFields.iobis
    }],
    emof
  }
}

function mapOccurrenceDataToTdwg({ basisOfRecord, beginDate, endDate, occurrenceStatus, scientificName, lifeStage, sex }) {
  return {
    basisOfRecord:    basisOfRecord.charAt(0).toUpperCase() + basisOfRecord.slice(1),
    eventDate:        format(beginDate, 'YYYY-MM-DD') + '/' + format(endDate || beginDate, 'YYYY-MM-DD'),
    occurrenceStatus: occurrenceStatus,
    scientificName:   scientificName,
    ...(lifeStage === 'unspecified' ? {} : { lifestage: lifeStage }),
    ...(sex === 'unspecified' ? {} : { sex })
  }
}

function mapDarwinCoreFieldsToRequest(darwinCoreFields) {
  return darwinCoreFields.reduce((acc, { name, value }) => {
    if (name.startsWith(purlUrl)) {
      acc.purl[name.substring(purlUrl.length)] = value
    } else if (name.startsWith(tdwgUrl)) {
      acc.tdwg[name.substring(tdwgUrl.length)] = value
    } else if (name.startsWith(iobisUrl)) {
      acc.iobis[name.substring(iobisUrl.length)] = value
    }
    return acc
  }, {
    purl:  {},
    tdwg:  {},
    iobis: {}
  })
}

export function mapDwcaToOccurrenceData(dwca) {
  const tdwg = dwca.dwcRecords.occurrence[0].tdwg
  const [beginDate, endDate] = tdwg.eventDate.split('/')
  return {
    basisOfRecord:    tdwg.basisOfRecord.charAt(0).toLowerCase() + tdwg.basisOfRecord.slice(1),
    beginDate:        new Date(beginDate),
    endDate:          endDate ? new Date(endDate) : null,
    lifestage:        tdwg.lifeStage || 'unspecified',
    occurrenceStatus: tdwg.occurrenceStatus,
    scientificName:   tdwg.scientificName,
    sex:              tdwg.sex || 'unspecified'
  }
}

export function mapDwcaToLocationData(dwca) {
  const tdwg = dwca.dwcRecords.occurrence[0].tdwg
  return {
    decimalLongitude:      tdwg.decimalLongitude,
    decimalLatitude:       tdwg.decimalLatitude,
    coordinateUncertainty: tdwg.coordinateUncertaintyInMeters,
    minimumDepth:          tdwg.minimumDepthInMeters,
    maximumDepth:          tdwg.maximumDepthInMeters,
    verbatimCoordinates:   tdwg.verbatimCoordinates,
    verbatimDepth:         tdwg.verbatimDepth
  }
}

export function mapDwcaToObservationData(dwca) {
  const tdwg = dwca.dwcRecords.occurrence[0].tdwg
  return {
    institutionCode:         tdwg.institutionCode,
    collectionCode:          tdwg.collectionCode,
    fieldNumber:             tdwg.fieldNumber,
    catalogNumber:           tdwg.catalogNumber,
    recordNumber:            tdwg.recordNumber,
    identifiedBy:            tdwg.identifiedBy.split('|'),
    recordedBy:              tdwg.recordedBy.split('|'),
    identificationQualifier: tdwg.identificationQualifier,
    identificationRemarks:   tdwg.identificationRemarks,
    references:              tdwg.associatedReferences.split('|')
  }
}

export function mapDwcaToMeasurements(dwca) {
  const emof = getProperty(() => dwca.dwcRecords.emof, [])
  return emof.map(({ tdwg, iobis }) => ({
    type:  tdwg.measurementType,
    value: tdwg.measurementValue,
    unit:  tdwg.measurementUnit,
    units: findUnitsByTypeId(iobis.measurementTypeID)
  }))
}

function getProperty(selectorFn, defaultValue) {
  try {
    const value = selectorFn()
    return value === null || value === undefined ? defaultValue : value
  } catch (e) {
    return defaultValue
  }
}

const reservedTdwgFields = ['basisOfRecord', 'eventDate', 'occurrenceStatus', 'scientificName', 'sex',
  'decimalLongitude', 'decimalLatitude', 'coordinateUncertaintyInMeters', 'minimumDepthInMeters', 'maximumDepthInMeters',
  'verbatimCoordinates', 'verbatimDepth', 'institutionCode', 'collectionCode', 'fieldNumber', 'catalogNumber',
  'recordNumber', 'identifiedBy', 'recordedBy', 'identificationQualifier', 'identificationRemarks',
  'associatedReferences']

export function mapDwcsToDarwinCoreFields(dwca) {
  const { tdwg, iobis, purl } = dwca.dwcRecords.occurrence[0]
  const iobisFields = Object
    .keys(iobis)
    .map(key => ({
      name:  iobisUrl + key,
      value: iobis[key]
    }))
  const purlFields = Object
    .keys(purl)
    .map(key => ({
      name:  purlUrl + key,
      value: purl[key]
    }))
  const tdwgFields = Object
    .keys(tdwg)
    .filter(key => !reservedTdwgFields.includes(key))
    .map(key => ({
      name:  tdwgUrl + key,
      value: tdwg[key]
    }))
  return [...iobisFields, ...purlFields, ...tdwgFields]
}
