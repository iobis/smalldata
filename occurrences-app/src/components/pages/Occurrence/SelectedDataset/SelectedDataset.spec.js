import React from 'react'
import renderer from 'react-test-renderer'
import SelectedDataset from './SelectedDataset'
import { getDatasetMock } from '../../../../clients/server'

describe('SelectedDataset', () => {
  it('renders correctly', () => {
    expect(renderer.create(
      <SelectedDataset
        datasets={getDatasetMock()}
        onChange={() => {}}
        selectedDataset={getDatasetMock()[0]}/>
    ).toJSON()).toMatchSnapshot()
  })
})
