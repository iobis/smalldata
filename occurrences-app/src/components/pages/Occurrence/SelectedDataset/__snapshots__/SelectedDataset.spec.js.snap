// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedDataset renders correctly 1`] = `
<div
  className="selected-dataset columns is-fluid"
>
  <table
    className="table is-striped is-fullwidth"
  >
    <tbody>
      <tr>
        <td>
          <input
            checked={true}
            name="dataset"
            onChange={[Function]}
            type="radio"
          />
        </td>
        <td>
          HAB Region 2: Occurrences of harmful (toxic) algal taxa within an area of interest to El Salvador compiled as part of a literature search project.
        </td>
      </tr>
      <tr>
        <td>
          <input
            checked={false}
            name="dataset"
            onChange={[Function]}
            type="radio"
          />
        </td>
        <td>
          NPPSD Short-tailed Albatross Sightings
        </td>
      </tr>
      <tr>
        <td>
          <input
            checked={false}
            name="dataset"
            onChange={[Function]}
            type="radio"
          />
        </td>
        <td>
          PANGAEA - Data from Christian-Albrechts-University Kiel
        </td>
      </tr>
      <tr>
        <td>
          <input
            checked={false}
            name="dataset"
            onChange={[Function]}
            type="radio"
          />
        </td>
        <td>
          NSIS: List of marine benthic algae from Magdalen Islands, Quebec as recorded in 1979
        </td>
      </tr>
      <tr>
        <td>
          <input
            checked={false}
            name="dataset"
            onChange={[Function]}
            type="radio"
          />
        </td>
        <td>
          Seguimiento de 10 cr√≠as de tortuga boba nacidas en 2016 en el litoral valenciano, en el marco del Proyecto LIFE 15 IPE ES 012 (aggregated per 1-degree cell)
        </td>
      </tr>
      <tr>
        <td>
          <input
            checked={false}
            name="dataset"
            onChange={[Function]}
            type="radio"
          />
        </td>
        <td>
          Waved Albatross Tracking (aggregated per 1-degree cell)
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
